### PR TITLE
Fix bad errors in the error list -- bug 1166689 

### PR DESCRIPTION
**Pull request by agocke**
This bug presents with bad errors in the error list when you reference a winmd project which itself references a winmd compilation. The fundamental issue causing the errors is that the compilation is not correctly unified with the assembly reference in the first winmd project. The cause of this is two different bugs: one in the project system and one in the compiler.

The issue with the compiler is that it does not unify compilations with output type winmdobj with references of ContentType WindowsRuntime because winmdobjs still have ContentType Default until they are rewritten by winmdexp. This is fixed in the second commit with a test to verify the unification.

The second issue is that the project system chooses the assembly name for a referenced compilation by looking at the output file name. The project system correctly strips off the .exe, .netmodule, and .dll extensions, but doesn't handle .winmdobj. A suggestion on how to test this change would be appreciated.

### CODE REVIEW

**Comment by agocke** [comment]
@jasonmalinowski FYI.

**Comment by agocke** [comment]
For the first commit, @tmat has mentioned that winmdobj's actually have ContentType default -- winmdexp must flip that bit when it runs over a winmdobj to turn it into a winmd. We've talked offline and agreed that a better solution is to special case this in the reference manager. I'm looking into it now.

**Comment by heejaechang** [comment]
retest this please

**Comment by jasonmalinowski** [comment]
:+1: for the IDE fix at least.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs [hunk 1]
```diff
@@ -1856,12 +1856,16 @@ internal override bool AreInternalsVisibleToThisAssembly(AssemblySymbol potentia
 
         private AssemblyIdentity ComputeIdentity()
         {
+            this.IsCompilationOutputWinMdObj();
```
Comment (Pilchie) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs]: This seems extraneous, given it's called again inline below.

File: src/Compilers/VisualBasic/Test/Semantic/Compilation/CompilationAPITests.vb [hunk 2]
```diff
@@ -21,6 +22,14 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.UnitTests
     Public Class CompilationAPITests
         Inherits BasicTestBase
 
+        <WorkItem(1166689, "DevDiv")>
+        <Fact>
+        Public Sub WinMDCompilationIdentity()
+            Assert.Equal(
+                AssemblyContentType.WindowsRuntime,
+                CreateCompilationWithMscorlib({"class C {}"}, options:=TestOptions.ReleaseWinMD).Assembly.Identity.ContentType)
```
Comment (Pilchie) [inline src/Compilers/VisualBasic/Test/Semantic/Compilation/CompilationAPITests.vb]: Why `DebugWinMD` in C# and `ReleaseWinMD` in VB?

**Comment by Pilchie** [comment]
LGTM

**Comment by tmeschter** [comment]
@dotnet-bot Retest this, please.

**Comment by agocke** [comment]
I've attempted to redo the compiler fix here and rewrite the commits -- we'll see if Git fully cooperated with me.

**Comment by agocke** [comment]
@tmat  Would you like to check the new change?

**Comment by agocke** [comment]
@jaredpar for shiproom

Bug 1166689
Scenario: This affects any WinRT solution which contains a reference to a winmd assembly that references a project in the solution or, more likely, a reference to a project of a non-C#/VB language that references another C#/VB project in the solution. So, for example, the simple solution of 1 C# project referencing a C++ WinRT project and a second C# project, where the C++ WinRT project also references the second C# WinRT project, will produce this issue. The issue manifests as all types being referenced by the C++ project as "missing" in the error list, producing errors. These are false errors and will not appear when the solution is actually built.

Testing done: Manual, to verify that the problem is fixed in the IDE, and unit testing for the compiler change.

**Comment by MattGertz** [comment]
Approved 6/12 by ML Shiproom.

**Pull request merged**